servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,231 @@
+import com.google.common.collect.Sets;
+
+import tachyon.Constants;
+
+/**
+ * A set of objects that can be indexed by specific fields of the object, and different IndexedSet
+ * instances can use different fields to index. The field type must be comparable. The field value
+ * must not be changed after being added to the set, otherwise, behavior for all operations is not
+ * specified.
+ *
+ * In operations that need field name as the parameter, if the object does not have the field, a
+ * RuntimeException will be thrown.
+ *
+ * This class is thread safe.
+ */
+public class IndexedSet<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** All objects in the set */
+  private final Set<T> mObjects;
+  /** Map from field index to an index of field related object in the internal lists */
+  private final Map<FieldIndex<T>, Integer> mIndexMap;
+  /** List of maps from fieldValue to set of T which is indexed by field name */
+  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** Final object for synchronization */
+  private final Object mLock = new Object();
+
+  /**
+   * A wrapper around a field name string to force the user to predefine the indexes, the user can
+   * reuse the instance of this class as parameters later other than directly use the error prone
+   * raw Strings.
+   */
+  public static interface FieldIndex<T> {
+    Object getFieldValue(T o);
+  }
+
+  /**
+   * Construct a new IndexedSet with at least one field as the index, the field can be either public
+   * or private. The {@link Field}s for these fields aren't retrieved until {@link #add(Object)} is
+   * called for the first time, it will be cached once it is retrieved so that reflection will only
+   * be needed once for a field.
+   *
+   * @param field at least one field is needed to index the set of objects
+   * @param otherFields other fields to index the set
+   */
+  public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
+    mObjects = new HashSet<T>();
+
+    mIndexMap = new HashMap<FieldIndex<T>, Integer>(otherFields.length + 1);
+    mIndexMap.put(field, 0);
+    for (int i = 1; i <= otherFields.length; i ++) {
+      mIndexMap.put(otherFields[i - 1], i);
+    }
+
+    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
+    for (int i = 0; i < mIndexMap.size(); i ++) {
+      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+    }
+  }
+
+  /**
+   * Add an object o to the set if there is no other object o2 such that (o == null ? o2 == null :
+   * o.equals(o2)). If this set already contains the object, the call leaves the set unchanged.
+   *
+   * @param object the object to add
+   * @return true if the object is successfully added to all indexes, otherwise false
+   */
+  public boolean add(T object) {
+    synchronized (mLock) {
+      boolean success = mObjects.add(object);
+      for (FieldIndex<T> field : mIndexMap.keySet()) {
```

Comment:
You eventually use the key and value of the mIndexMap. Maybe you should just iterate over the entrySet().